gpt4all-chat/database.cpp

```diff
@@ -442,11 +576,84 @@ bool selectDocuments(QSqlQuery &q, int folder_id, QList<int> *documentIds) {
+    if (!q.exec())
+        return false;
+    if (q.next()) {
+        *total_docs = q.value(0).toInt();
+        *total_words = q.value(1).toInt();
+        *total_tokens = q.value(2).toInt();
+    }
+    return true;
+}
+
+QSqlError Database::initDb()
 {
-    QString dbPath = MySettings::globalInstance()->modelPath()
-        + QString("localdocs_v%1.db").arg(LOCALDOCS_VERSION);
+    /*
+     * Support upgrade path from older versions:
+     *
+     *  1. Detect and load dbPath with older versions
+     *  2. Provide versioned SQL select statements
+     *  3. By default mark all collections of older versions as force indexing and present to the user
+     *     the an 'update' button letting them know a breaking change happened and that the collection
+     *     will need to be indexed again
+     *  4. Upgrade the tables to the new version
+     *  5. For some version upgrades we may be able to write bespoke code that does not require a
+     *     forced indexing of older collections, but in lieu of specific versioned upgrades the default
+     *     will require forced indexing to be safe
+     */
+
+    // Iterate through the files and find the one with the largest version number
+    QDirIterator dbIt(MySettings::globalInstance()->modelPath(), QDir::Files);
+    QRegularExpression regex("localdocs_v(\\d+)\\.db");
+    QRegularExpressionMatch match;
+
+    QString filename;
+    int version = 0;
+
+    while (dbIt.hasNext()) {
+        dbIt.next();
+        QString currentFilename = dbIt.fileName();
+        match = regex.match(currentFilename);
+        if (match.hasMatch()) {
+            int versionNumber = match.captured(1).toInt();
+            if (versionNumber > version) {
+                version = versionNumber;
+                filename = currentFilename;
+            }
+        }
+    }
+
+    // If we're upgrading, then we need to do a select on the current version of the collections table,
+    // then delete the current db file, create the new one and populate the collections table and mark
+    // them as needing forced indexing
+
+    QList<CollectionItem> collections;
     QSqlDatabase db = QSqlDatabase::addDatabase("QSQLITE");
+    if (version != LOCALDOCS_VERSION) {
+#if defined(DEBUG)
+        qDebug() << "Older localdocs version found" << version << "upgrade to" << LOCALDOCS_VERSION;
+#endif
+        const QString oldDatabasePath = MySettings::globalInstance()->modelPath() + filename;
+        db.setDatabaseName(oldDatabasePath);
+
+        if (!db.open())
+            qWarning() << "ERROR: Could not open old db file" << db.lastError();
+
+        // Select the current collections which will be marked to force indexing
+        QSqlQuery q;
+        if (!selectAllFromCollections(version, q, &collections))
+            qWarning() << "ERROR: Could not open select old collections" << q.lastError();
+
+        // Remove the old database
+        db.close();
+        QFile::remove(oldDatabasePath);
```

Comment:
The entire database including the tables and structure is being **replaced** here with the only params we need held in memory: the collection paths and their names. There is zero reason to use a transaction to do this as a file system transaction will do.

Reply:
But a transactional filesystem operation would look like:
- Read the old database
- Save the new database
- flush any write buffers maintained by Qt or SQLite
- `fsync()` the new database (ensuring it is flushed to physical storage, using F_FULLFSYNC on macOS)
- `fdatasync()` the model folder
- THEN, you can remove the old database.

And to ensure atomicity, the code must do one of the following:
  - Use the oldest available database, and assume the update was incomplete if two versions are present
  - Use the same filename for both db versions, and rely on the atomicity of rename() (seems like it's too late to go this route now)
  - Use some flag in the database to mark that the database has completely copied and synced to disk


What can happen with the current code:
- Old database is read
- Old database is removed
- (Optional) New database begins to be written to disk
- GPT4All is forcefully terminated, system crash, power failure, etc.
- Next launch of GPT4All will either have no database, or a partially written new database. This is something SQLite is designed to avoid, and we are losing this guarantee as this PR is currently written.